src/compiler_lectures/how_salsa_works.md

```diff
@@ -0,0 +1,169 @@
+
+## How does it work?
+
+The first thing that Salsa has to do is identify the "base inputs" [^EN1].
+
+Then Salsa has to also identify intermidiate, "derived" values, which are something that the library produces, but, for each derived value there's a "pure" function that computes the derived value.
+
+For example, there might be a function `ast(x: Path) -> AST`. The produced `AST` isn't a final value, it's an intermidiate value that the library would use for the computation.
+
+This means that when you try to compute with the library, Salsa is going to compute various derived values, and eventually read the input and produce the result for the asked computation.
+
+Salsa is going to track, in the course of computing, which inputs were accessed, which derived values, and this is going to be used later to determine what's going to happen when the inputs change: are the derived values still valid? 
+
+This doesn't mean necessarily that each computation downstream from the input is going to be checked, since that is going to be costly. Salsa only needs to check each downstream computation until it finds one that isn't changed, therefore it won't need to check the other derived computations, since they wouldn't need to change.
+
+It's is helpful to think about this as a graph with nodes. Each derived value has a dependency on some other values, that could be base or derived. Base values don't have a dependency.
+
+```ignore
+I <- A <- C ...
+          |
+J <- B <--+
+```
+
+When an input `I` changes, the derived value `A` could have changed, but another derived value `B` , which doesn't depend neither on `I`, nor does it depends on `A`, nor does it depends on any other derived value from `A` or `I`, is not subject to change.
+Therefore, Salsa can reuse the computation done for `B` in the past, without having to compute it again.
+
+The computation could also terminate early. Keeping the same graph as before, say that input `I` has changed in some way (and input `J` hasn't) but, when computing `A` again, it's found that `A` hasn't changed from the previous computation. This leads to an "early termination", because there's no need to check if `C` needs to change, since both `C` direct inputs, `A` and `B`, haven't changed.
+
+## Key Salsa concepts
+### Query
+A query is some value that Salsa can access in the course of computation.
+Each query can have a number of keys (from 0 to many), and all queries have a result, akin to functions.
+0-key queries are called "input" queries.
+### Database
+The database is basically the context for the entire computation, it's gonna store all Salsa's internal state, all intermidiate values for each query, and anything else that the computation might need.
+The database to know all the queries that the library is going to do before it can be build, but they don't need to be specified in the same place.
```

Comment:
```suggestion
The database must know all the queries that the library is going to do before it can be built, but they don't need to be specified in the same place.
```